File: Problem2.py
#Did this code successfully run on Leetcode : Yes
#Any problem you faced while coding this : No

class WordDistance:
    #Time Complexity :O(n)
    #Space Complexity :O(n)
    def __init__(self, wordsDict: List[str]):
        self.map={}
        for i in range(len(wordsDict)):
            if wordsDict[i] not in self.map.keys():
                self.map[wordsDict[i]]=[]
            self.map[wordsDict[i]].append(i)
            
    #Time Complexity :O(k1+k2) where k1=size of l1  and k2=size of l2, worst case O(n)
    #Space Complexity :O(1)
    def shortest(self, word1: str, word2: str) -> int:
        li1=self.map.get(word1)
        li2=self.map.get(word2)
        p1=0
        p2=0
        minVal=float('inf')
        while p1<len(li1) and p2<len(li2):
            if li1[p1]<li2[p2]:
                minVal=min(minVal,li2[p2]-li1[p1])
                p1+=1
            else:
                minVal=min(minVal,li1[p1]-li2[p2])
                p2+=1
        return minVal


# Your WordDistance object will be instantiated and called as such:
# obj = WordDistance(wordsDict)
# param_1 = obj.shortest(word1,word2)

File: problem2.cpp
/*
// Time Complexity : O(N) for each call
// Space Complexity : O(N) number of words.
// Did this code successfully run on Leetcode :
// Any problem you faced while coding this :


// Your code here along with comments explaining your approach
*/

#include<iostream>
#include<unordered_map>
#include<vector>
#include<string>

using namespace std;

class WordDistance {
public:
    unordered_map<string,vector<int>> umap{};
    size_t max_size{};
    WordDistance(vector<string>& wordsDict) {
        this->max_size = wordsDict.size();
        for(int i{};i<this->max_size;++i){
            string word = wordsDict.at(i);
            if(umap.find(word) == umap.end()){
                this->umap[word] = vector<int>{};
            }
            this->umap[word].push_back(i);
        }
    }
    
    int shortest(string word1, string word2) {
        int i{},j{};
        size_t w1_sz{},w2_sz{};
        w1_sz = this->umap[word1].size();
        w2_sz = this->umap[word2].size();
        int max_diff = this->max_size;
        while(i<w1_sz && j<w2_sz){
            int val = abs(this->umap[word1].at(i) - this->umap[word2].at(j));
            max_diff = min(max_diff,val);
            if(this->umap[word1].at(i)<this->umap[word2].at(j)){
                ++i;
            }
            else if(this->umap[word1].at(i)>this->umap[word2].at(j)){
                ++j;
            }
        }
        while(i<w1_sz){
            int val = abs(this->umap[word1].at(i)-this->umap[word2].at(j-1));
            max_diff = min(max_diff,val);
            ++i;
        }
        while(j<w2_sz){
            int val = abs(this->umap[word1].at(i-1)-this->umap[word2].at(j));
            max_diff = min(max_diff,val);
            ++j;
        }
        return max_diff;
    }
};

/**
 * Your WordDistance object will be instantiated and called as such:
 * WordDistance* obj = new WordDistance(wordsDict);
 * int param_1 = obj->shortest(word1,word2);
 */